### DISCUSSION TITLE
No Rendered Tiles

### DISCUSSION THREAD

**Discussion by helloimalastair** [maplibre/maplibre-gl-js#2239] [Q&A] [answered]
Hey, sorry for the necro on [this discussion](https://github.com/maplibre/maplibre-gl-js/discussions/2191). I tried to redo [the HTML](https://github.com/maplibre/maplibre-gl-js/files/10777249/test_goalastair.zip) for my map(provided by *zerda-ocm*) and now it seems to be pulling the `.pmtiles` correctly, but again nothing renders.

I basically just hot-swapped the setup that worked with the smaller [`map.pmtiles`](https://maps.goalastair.com/map.pmtiles) for the larger [`planet.pmtiles`](https://maps.goalastair.com/planet.pmtiles).

The `.pmtiles` does appear to be functional, as on the [PMTiles Viewer](https://protomaps.github.io/PMTiles/?url=https%3A%2F%2Fmaps.goalastair.com%2Fplanet.pmtiles#map=0.66/0/61.5) it shows up, with labels too, so I'm kind of stumped what else I could be missing...

This would be a lot easier to debug myself, if I saw any errors, but the only thing I can see is that the `.pmtiles` chunks are sometimes *slightly* slow to load. I've also tried panning around the map, to no avail. Is there some kind of debug flag I can enable to get more info on what I might be doing wrong?

**Comment by zerda-ocm** [comment, marked as answer]
The process you used to generate the planet.pmtiles is different to the one you used for map.pmtiles.
For map.pmtiles you were using the openmaptiles schema which organises and processes features in different layers. For example buildings are in the layer _building_ and landuses are in the layer _landuse_ (additionally changing _landuse="meadow"_ to _class="grass"_, etc,).
In your planet.pmtiles on the other hand, all data is shoved in one single layer "_osm_" and the orginal osm tags are used. (activate "_show attributes_" in the pmtiles viewer, than it should be become clearer, that you are using a completely different data structure) 

Your style is expecting _layer="landuse"_ and _class="grass"_, but your data only has _layer="osm"_ and _landuse="meadow"_

So you either have to alter your style or use the same process to generate the planet.pmtiles as you have used for your map.pmtiles.

**Reply by helloimalastair** [reply]
Oh, that's kind of irritating. I was using a prebuilt `.mbtiles` file from Mapbox, but if I have to generate it all myself, it might take a while. Thanks for the input!

**Reply by zerda-ocm** [reply]
Tilemaker is not really suitable for a full planet import. Take a look at [planetiler](https://github.com/onthegomap/planetiler).

**Reply by helloimalastair** [reply]
I'll retry when I get back, but I remember trying that one, and I could never get it to start downloading the entire planet. It would always default to Monaco, unless I explicitly set another region with `--area`(as in, if it was omitted, it would default to Monaco).

**Reply by zerda-ocm** [reply]
This is normal behaviour and --area=planet will download the entire planet. See https://github.com/onthegomap/planetiler/blob/main/PLANET.md for some tweaks in regards of the specification of your system.

And since your current style is based on openmaptiles you might want to a look at https://github.com/openmaptiles/planetiler-openmaptiles too.

**Reply by helloimalastair** [reply]
Ooh, nice, thanks for the recommendation, running that now!

**Reply by helloimalastair** [reply]
Just wondering, how long is it supposed to take for a full generation? I've been running for about 18 hours now, and one of the meters is still only at 2%. As far as I can tell, it isn't downloading anything right now(the biggest bottleneck I expected), so I'm at a loss. I'm running on a 12-core Ryzen 5, with 16 GiB of RAM + 40 GiB of Swap, and the utilization on both RAM and CPU is low, so I'm not sure what else might be causing it...

Progress, for reference:
```
18:32:21 INF [osm_pass2] -  nodes: [ 8.2B 100%    0/s ] 75G   ways: [  17M   2%  600/s ] rels: [    0   0%    0/s ] features: [ 641M 2.3k/s ] 51G   blocks: [  26k  66%    0/s ]
    cpus: 0 gc:  0% heap: 2.9G/4.1G direct: 18M postGC: 1.9G relInfo: 520M mpGeoms: 248M 
    read( 0%) ->   (11/22) -> process( 1%  0%  0%  0%  1%  0%  0%  0%  0%  0%  0%) -> (4.6k/62k) -> write( 0%)
```

**Reply by zerda-ocm** [reply]
The biggest bottleneck is not the download (you have done that already), but the IO-performance of the SSD/HDD.
When utilization of RAM and CPU is both low, it is very likely that your HDD/SSD is at 100% (I use command `atop` + button `D` to monitor the disk performance).
For reference: I am using a Strato VSERVER with 60GB RAM and if i remember correctly the SSD performance was only around 100 MB/s. A full planet import took around 3 days. With better Hardware (lots of RAM and very fast SSD) a planet import can be done in a fraction of the time: https://github.com/onthegomap/planetiler#benchmarks

**Reply by helloimalastair** [reply]
Oh yikes. I'm running at 100%, so maybe that is why. Any general tips on seeing how far along I am? The 2% indicator for ways is worrying, so I'm hoping it will speed up at some point, if only just a little...

**Reply by zerda-ocm** [reply]
What are your approx. stats on RDDSK and WRDSK?

**Reply by helloimalastair** [reply]
Here's the entire row from `atop`:
```sh
DSK |          sda | busy    100%  | read    1806 | write     10 | discrd     0  | KiB/r      4 | KiB/w    123 |              |  KiB/d      0 | MBr/s    0.7 | MBw/s    0.1 |  avq    10.13 | avio 5.50 ms |
```

**Reply by zerda-ocm** [reply]
Judging by the speed, you are using a HDD and not a SSD. In other words: This will probably take weeks to finish. You either have to upgrade your system with a decent SSD or you can have a look into cloud computing.

**Reply by helloimalastair** [reply]
Oh, ok. Yeah, I have an SSD on that system, but I don't know if it will fit, as it already has a dual install of Windows/Ubuntu.

How about an Oracle Free Tier VPS? Would that be good enough? Specs:
```
2 AMD based Compute VMs with 1/8 OCPU** and 1 GB memory each

Arm-based Ampere A1 cores and 24 GB of memory usable as 1 VM or up to 4 VMs with 3,000 OCPU hours and 18,000 GB hours per month
```

**Reply by zerda-ocm** [reply]
Hmm, for a SSD that is surprisingly low. You can try `iotop `to have a better understanding which processes are the bottleneck here.

Regarding Oracle Free Tier VPS: I have no idea. https://github.com/onthegomap/planetiler#benchmarks lists some typical cloud services. I haven't tried it yet, but spot prices for AWS can be pretty inexpensive (c6gd.4xlarge for around $0.16 an hour).

**Reply by helloimalastair** [reply]
Sorry, I didn't write that out correctly. I meant that my system has an NVME SSD as the system drive, but at only 500 GB, I'm hesitant to try putting `planetiler` onto it, for fear of accidentally overrunning the storage I have. This is why I was running it on my admittedly way too slow 2TB HDD.

Thanks for the recommendations! On the `c6gd.4xlarge` front, how many hours would you say it would take to complete with a VPS like that?

**Reply by wipfli** [reply]
In the planetiler slack channel in https://slack.openstreetmap.us someone reported this:

> I built Europe using the Oracle Cloud Free Tier in 6h54m40s. You get an ARM Server with 4 vCPUs, 24GB RAM and 200GB disk.
 docker run -e JAVA_TOOL_OPTIONS="-Xmx14g" -v "$(pwd)/data":/data ghcr.io/onthegomap/planetiler:latest --download --area=europe

**Reply by wipfli** [reply]
But yeah, keep in mind that this is MapLibre GL JS and probably it is best to discuss planetiler settings either in the planetiler github repo or the planetiler slack.

**Reply by zerda-ocm** [reply]
On a c6gd.4xlarge around 3 hours, but you will achieve the same result with your SSD.

**Reply by helloimalastair** [reply]
@wipfli, oh oops, sorry about that! And thanks for the help, both of you!
